### PR TITLE
[IR] Add Support for Struct Types

### PR DESCRIPTION
**Pull request by zzzDavid**
## Description ##
This PR introduces support for struct types in the Allo. It adds the ability to define, manipulate, and resolve struct types in the IR, enabling more expressive representations and operations in programs. The implementation resolves dictionary literals to structs and provides infrastructure for struct construction, field access, and type inference.

### Problems ###
- Lack of support for struct types in the Allo frontend.

### Proposed Solutions ###
- Added struct type handling across the compiler pipeline:
  - **IR/Builder**: Added methods for struct construction (`StructConstructOp`) and field access (`StructGetOp`).
  - **Type Inference**: Enhanced type inference for struct types and dictionary literals, mapping keys to their respective types.
  - **Symbol Resolver**: Resolved dictionary literals into struct types during AST traversal.
  - **IR Transformations**: Implemented lowering for composite (struct) types.
- Introduced corresponding unit tests to validate the struct implementation.
- Enhanced `test_types.py` to include a comprehensive test case for a simple struct-based kernel.

#### Input Program ####
```python
struct_type = T.Struct({"x": int32, "y": float32})

def kernel(x: int32[16], y: float32[16]) -> int32:
    sum_val: int32 = 0
    for i in range(16):
        point: struct_type = {"x": x[i], "y": y[i]}
        sum_val += point["x"]
        sum_val += int(point["y"])
    return sum_val
```

### Expected Behavior

- Generated IR:
    - Struct construction (StructConstructOp) for creating the point variable.
    - Field access (StructGetOp) for extracting "x" and "y" fields from point.
- Execution: Outputs correct sum of integer and float elements.

### Checklist

- [x] PR's title starts with a category (e.g. [IR], [Struct], [TypeSystem]).
- [x] Changes are complete (i.e., coding and testing for this PR are finished).
- [x] All changes have test coverage (unit tests included for struct handling and edge cases).
- [x] Code is well-documented.

### CODE REVIEW

File: allo/ir/infer.py [hunk 1]
```diff
@@ -396,9 +410,24 @@ def visit_symbol(ctx, node):
         # pylint: disable=raising-bad-type
         raise None
 
+    @staticmethod
     @staticmethod
```
Comment (chhzh123) [inline allo/ir/infer.py]: Two `@staticmethod`

**Review by chhzh123 — COMMENTED** [review]
Mostly LGTM

File: allo/ir/infer.py (hunk 1 above)
Comment (zzzDavid) [reply allo/ir/infer.py]: Thanks, I missed that. The PR will be ready soon after I fix a test

**Comment by chhzh123** [comment]
Please let me know if you are ready to merge @zzzDavid

**Comment by zzzDavid** [comment]
Yes, it's ready
